// @dojo/framework/core/animations/cssTransitions.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    var browserSpecificTransitionEndEventName = '';
    var browserSpecificAnimationEndEventName = '';
    function determineBrowserStyleNames(element) {
        if ('WebkitTransition' in element.style) {
            browserSpecificTransitionEndEventName = 'webkitTransitionEnd';
            browserSpecificAnimationEndEventName = 'webkitAnimationEnd';
        }
        else if ('transition' in element.style || 'MozTransition' in element.style) {
            browserSpecificTransitionEndEventName = 'transitionend';
            browserSpecificAnimationEndEventName = 'animationend';
        }
        else {
            throw new Error('Your browser is not supported');
        }
    }
    function initialize(element) {
        if (browserSpecificAnimationEndEventName === '') {
            determineBrowserStyleNames(element);
        }
    }
    function runAndCleanUp(element, startAnimation, finishAnimation) {
        initialize(element);
        var finished = false;
        var transitionEnd = function () {
            if (!finished) {
                finished = true;
                element.removeEventListener(browserSpecificTransitionEndEventName, transitionEnd);
                element.removeEventListener(browserSpecificAnimationEndEventName, transitionEnd);
                finishAnimation();
            }
        };
        startAnimation();
        element.addEventListener(browserSpecificAnimationEndEventName, transitionEnd);
        element.addEventListener(browserSpecificTransitionEndEventName, transitionEnd);
    }
    function exit(node, exitAnimation, active) {
        var exitAnimationClasses = exitAnimation.split(' ');
        var activeClasses = active && active !== true ? active.split(' ') : exitAnimationClasses.map(function (className) { return className + "-active"; });
        runAndCleanUp(node, function () {
            exitAnimationClasses.forEach(function (className) { return node.classList.add(className); });
            requestAnimationFrame(function () {
                activeClasses.forEach(function (className) { return node.classList.add(className); });
            });
        }, function () {
            node && node.parentNode && node.parentNode.removeChild(node);
        });
    }
    function enter(node, enterAnimation, active) {
        var enterAnimationClasses = enterAnimation.split(' ');
        var activeClasses = active && active !== true ? active.split(' ') : enterAnimationClasses.map(function (className) { return className + "-active"; });
        runAndCleanUp(node, function () {
            enterAnimationClasses.forEach(function (className) { return node.classList.add(className); });
            requestAnimationFrame(function () {
                activeClasses.forEach(function (className) { return node.classList.add(className); });
            });
        }, function () {
            enterAnimationClasses.forEach(function (className) { return node.classList.remove(className); });
            activeClasses.forEach(function (className) { return node.classList.remove(className); });
        });
    }
    exports.default = {
        enter: enter,
        exit: exit
    };
});
